hammer/sim/xcelium/__init__.py

```diff
@@ -458,4 +556,337 @@ def sim_xrun(self) -> bool:
+      """
+      ### Read in disciplines file, if it exists.
+      
+  
+      dpath = os.path.join(os.getcwd(), discipline_filename)
+
+      if not os.path.isfile(dpath):
+        self.logger.error(f"No discipline file found at {dpath}.")
+
+      if dpath:
+        df = open(dpath, "r")
+        discipline_opts = df.read()
+        disciplines_formatted = re.sub("\n", " \\\n", discipline_opts) + " \\"
+        df.close()
+        return disciplines_formatted
+      else:
+        return ""
+
+  def option_preparer(self, opts, addt_opts, pseudo_step=True) -> str:
+      """
+      Returns a formatted string of all provided AMS options and their arguments
+      """
+      bool_list = ["ams", "disciplines", "gen_amscf"]
+      opts_proc = opts.copy()
+      header = ""
+
+      if not opts:
+        return ""
+
+      if opts ["ams"] is True:
+          opts_proc ["ams"] = "-ams_flex" + " \\"
+      else:
+        opts_proc ["ams"] = ""
+      
+      if opts ["gen_amscf"] is True:
+        self.generate_amscf(self.get_setting("sim.xcelium.amscf_template"), self.get_setting("sim.xcelium.amscf")) # Expect names, not filepaths
+        opts_proc ["gen_amscf"] = ""
+      else:
+        opts_proc ["gen_amscf"] = ""
+
+      if opts ["disciplines"]:
+        header += self.discipline_collector(opts["disciplines"]) + "\n"
+      
+      if opts ["amsconnrules"]:
+        opts_proc ["amsconnrules"] = opts["amsconnrules"]
+
+      if (pseudo_step):
+          digital_opts = self.option_extractor(["xrun_compile.arg", "xrun_elab.arg", "xrun_sim.arg"])
+          digital_opts.update(opts_proc) # Should any keys match, AMS arguments take precedence
+          opts_proc = digital_opts
+
+      opts_proc = {opt:setting for (opt, setting) in opts_proc.items() if opt not in bool_list and setting is not None}
+
+      opts_len = len(opts_proc) - 1
+      for (n, (opt, setting)) in enumerate(opts_proc.items()):
+        if (n == opts_len):
+            if (setting == ""):
+                opts_proc [opt] = f"-{opt}"
+            else:
+                opts_proc [opt] = f"-{opt} {setting}"
+        else:
+          if (setting == ""):
+              opts_proc [opt] = f"-{opt} \\"
+          else:
+              opts_proc [opt] = f"-{opt} {setting} \\"
+
+
+      opts_rev = {k: v for k, v in opts_proc.items() if v}
+
+      opts_proc_str = "\n".join(opts_rev.values())
+
+      # Attach user-defined commands, if any are included
+      if (addt_opts):
+        opts_proc_str += " \\\n"
+
+      footer = " \\\n".join(addt_opts)
+
+      return header + f"{opts_proc_str}" + footer
+
+  def generate_amscf(self, template_filename, amscontrol_filename) -> bool:
+      """
+      Creates an AMS control file based on templated format with available analog models & schematics
+      """
+      
+      # Get analog models, schematics from directories specified in extralibs
+      extralib = self.get_setting("vlsi.technologies.extra_libraries")
```

Comment:
This key doesn't exist... what are you trying to do/specify here? Do you need fields added to the normal `vlsi.technology.extra_libraries` key? https://github.com/ucb-bar/hammer/blob/5277152e35f664a23e816c26fcc66e605173b19a/hammer/tech/__init__.py#L249